src/Compilers/CSharp/Test/Emit3/Semantics/CollectionExpressionTests_WithElement_Constructor.cs

```diff
@@ -0,0 +1,1365 @@
+    public void WithElement_NestedType()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class Outer
+            {
+                public class MyList<T> : List<T>
+                {
+                    public int Value { get; }
+                    
+                    public MyList(int value) : base()
+                    {
+                        Value = value;
+                    }
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    Outer.MyList<int> list = [with(42), 1];
+                    Console.WriteLine(list.Value);
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("42"));
+    }
+
+    #endregion
+
+    #region Complex Scenarios
+
+    [Fact]
+    public void WithElement_WithExpressionInArguments()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public int Value { get; }
+                
+                public MyList(int value) : base()
+                {
+                    Value = value;
+                }
+            }
+            
+            class C
+            {
+                static void Main()
+                {
+                    int x = 10;
+                    MyList<int> list = [with(x + 32), 1];
+                    Console.WriteLine(list.Value);
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("42"));
+    }
+
+    [Fact]
+    public void WithElement_WithMethodCall()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public string Value { get; }
+                
+                public MyList(string value) : base()
+                {
+                    Value = value;
+                }
+            }
+            
+            class C
+            {
+                static string GetValue() => "test";
+                
+                static void Main()
+                {
+                    MyList<int> list = [with(GetValue()), 1];
+                    Console.WriteLine(list.Value);
+                }
+            }
+            """;
+
+        CompileAndVerify(source, expectedOutput: IncludeExpectedOutput("test"));
+    }
+
+    [Fact]
+    public void WithElement_WithLambda()
+    {
+        var source = """
+            using System;
+            using System.Collections.Generic;
+            
+            class MyList<T> : List<T>
+            {
+                public Func<int> ValueFunc { get; }
+                
+                public MyList(Func<int> func) : base()
```

Comment:
Consider testing with a `Delegate`-taking constructor to involve lambda natural typing.

Reply:
It would also be interesting to see some tests with deep inference involved. Consider:

```cs
M([with(() => 1), () => 2]);
void M<T>(MyList<T> list) { }
```

when `MyList` has a `T`-accepting constructor. Consider testing such a scenario in multiple ways, when the lambdas agree on their type, when they don't, and when there can be conversions from one return type to the other.

Reply:
b06589b6dd9 and b21acf28d1e